reward/staking_manager.go

```diff
@@ -47,17 +52,34 @@ type StakingManager struct {
 	blockchain           blockChain
 	chainHeadChan        chan blockchain.ChainHeadEvent
 	chainHeadSub         event.Subscription
+	migrationCheck       migrationCheck
 }
 
-func NewStakingManager(bc blockChain, gh governanceHelper, db stakingInfoDB) *StakingManager {
-	return &StakingManager{
+func NewStakingManager(bc blockChain, gh governanceHelper, db database.DBManager) *StakingManager {
+	sm := &StakingManager{
 		addressBookConnector: newAddressBookConnector(bc, gh),
 		stakingInfoCache:     newStakingInfoCache(),
-		stakingInfoDB:        db,
+		stakingInfoDB:        stakingInfoDB(db),
 		governanceHelper:     gh,
 		blockchain:           bc,
 		chainHeadChan:        make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
+		migrationCheck:       migrationCheck(db),
 	}
+
+	// Before migration, staking information of current and before should be stored in DB.
+	//
+	// Staking information from block of StakingUpdateInterval ahead is needed to create a block.
+	// If there is no staking info in either cache, db or state trie, the node cannot make a block.
+	// The information in state trie is deleted after state trie migration.
+	blockchain.RegisterMigrationPrerequisites(func(blockNum uint64) error {
+		if _, err := sm.updateStakingInfo(blockNum); err != nil {
+			return err
+		}
+		_, err := sm.updateStakingInfo(blockNum + params.StakingUpdateInterval())
```

Comment:
What if the current block number is less than `blockNum + params.StakingUpdateInterval()`?

Reply:
Sorry, I put `CalcStakingBlockNumber` in updateStakingInfo.
Now, current block number  can never be less than `blockNum + params.StakingUpdateInterval()`.

Reply:
Thanks to @KimKyungup I made a new function.